tests/Result.cpp

```diff
@@ -2,85 +2,112 @@
 
-        detail::Result<int> resultError(badResult);
-        CHECK(*resultError == 0);
-        CHECK(*std::as_const(resultError) == 0);
-
         SUBCASE("rvalue") {
-            CHECK(*detail::Result<int>(1) == 1);
-            CHECK(*detail::Result<int>(badResult) == 0);
+            CHECK(*Result<int>(1) == 1);
         }
     }
 
+    SUBCASE("code") {
+        CHECK(Result<int>(1).code() == UA_STATUSCODE_GOOD);
+        CHECK(Result<int>(badResult).code() == badCode);
+    }
+
+    SUBCASE("hasValue") {
+        Result<int> result(1);
+        CHECK(result.hasValue());
+
+        Result<int> resultError(badResult);
+        CHECK_FALSE(resultError.hasValue());
+    }
+
     SUBCASE("value") {
-        detail::Result<int> result(1);
+        Result<int> result(1);
         CHECK(result.value() == 1);
         CHECK(std::as_const(result).value() == 1);
 
-        detail::Result<int> resultError(badResult);
+        Result<int> resultError(badResult);
         CHECK_THROWS_AS(resultError.value(), BadStatus);
         CHECK_THROWS_AS(std::as_const(resultError).value(), BadStatus);
 
         SUBCASE("rvalue") {
-            CHECK(detail::Result<int>(1).value() == 1);
-            CHECK_THROWS_AS(detail::Result<int>(badResult).value(), BadStatus);
+            CHECK(Result<int>(1).value() == 1);
+            CHECK_THROWS_AS(Result<int>(badResult).value(), BadStatus);
         }
     }
 
     SUBCASE("valueOr") {
-        detail::Result<int> result(1);
+        Result<int> result(1);
         CHECK(result.valueOr(2) == 1);
 
-        detail::Result<int> resultError(badResult);
+        Result<int> resultError(badResult);
         CHECK(resultError.valueOr(2) == 2);
 
         SUBCASE("rvalue") {
-            CHECK(detail::Result<int>(1).valueOr(2) == 1);
-            CHECK(detail::Result<int>(badResult).valueOr(2) == 2);
+            CHECK(Result<int>(1).valueOr(2) == 1);
+            CHECK(Result<int>(badResult).valueOr(2) == 2);
         }
     }
 }
 
 TEST_CASE("Result (void template specialization)") {
+    SUBCASE("operator StatusCode") {
+        CHECK(static_cast<StatusCode>(Result<void>()) == UA_STATUSCODE_GOOD);
+        CHECK(static_cast<StatusCode>(Result<void>(badResult)) == UA_STATUSCODE_BADUNEXPECTEDERROR);
+    }
+
     SUBCASE("code") {
-        CHECK(detail::Result<void>().code() == UA_STATUSCODE_GOOD);
-        CHECK(detail::Result<void>(badResult).code() == badCode);
+        CHECK(Result<void>().code() == UA_STATUSCODE_GOOD);
+        CHECK(Result<void>(badResult).code() == badCode);
+    }
+
+    SUBCASE("value") {
+        Result<void> result{};
+        CHECK_NOTHROW(result.value());
+
+        Result<void> resultError(badResult);
+        CHECK_THROWS_AS(resultError.value(), BadStatus);
+
+        SUBCASE("rvalue") {
+            CHECK_NOTHROW(Result<void>().value());
+            CHECK_THROWS_AS(Result<void>(badResult).value(), BadStatus);
+        }
     }
 }
 
 TEST_CASE("tryInvoke") {
```

Comment:
Added a test for the implicit conversion to StatusCode.

I originally also had an implicit conversion to UA_StatusCode but had to drop it because of MSVC reporting some ambiguity errors. I have not checked further and thought it should be okay not to include it, because StatusCode itself already provides an implicit conversion to UA_StatusCode (though both conversions at once doesn't seem to work in one go).  Do you agree?

Reply:
I think is dangerous to add a template specialization for `UA_StatusCode` because it's just an alias  for `uint32_t`. Functions returning `uint32_t` would be captured as `Result<void>`.

Reply:
Changed tthe template specialization from UA_StatusCode to BadResult and added a test for it. I'm not sure if we really need it, because a function always returning BadResult would be quite odd. On the other hand I guess it doesn't hurt 😊